infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuInterClusterReplicationServer.java

```diff
@@ -290,21 +306,75 @@ private ServerContext getServerContext(Map<String, Object> opts) {
     }
 
     /**
-     * Start Corfu Log Replication Discovery Service
+     * Start Corfu Log Replication Discovery Service after the local node version matches the cluster version, i.e.,
+     * any ongoing rolling upgrade completes.
      *
      * @param serverContext server context (server information)
      */
     private void startDiscoveryService(ServerContext serverContext) {
 
+        CorfuRuntime runtime = getRuntime(serverContext);
+        CorfuStore corfuStore = new CorfuStore(runtime);
+
+        LogReplicationUpgradeManager upgradeManager =
+                new LogReplicationUpgradeManager(corfuStore, serverContext.getPluginConfigFilePath());
+
+        // Check if an upgrade is in progress.  If it is, wait for it to complete
+        log.info("Wait for any ongoing rolling upgrade to complete.");
+        while(isUpgradeInProgress(upgradeManager, corfuStore)) {
+            // Wait before checking again.
+            sleep();
+        }
+
+        // Upgrade has completed.  Start the discovery Service
         log.info("Start Discovery Service.");
 
         // Start LogReplicationDiscovery Service, responsible for
         // acquiring lock, retrieving Site Manager Info and processing this info
         // so this node is initialized as Source (sender) or Sink (receiver)
-        replicationDiscoveryService = new CorfuReplicationDiscoveryService(serverContext);
+        replicationDiscoveryService = new CorfuReplicationDiscoveryService(serverContext, runtime);
         replicationDiscoveryService.start();
     }
 
+    private CorfuRuntime getRuntime(ServerContext serverContext) {
+
+        String localCorfuEndpoint = getCorfuEndpoint(getHostFromEndpointURL(serverContext.getLocalEndpoint()),
+            serverContext.getCorfuServerConnectionPort());
+
+        return CorfuRuntime.fromParameters(CorfuRuntime.CorfuRuntimeParameters.builder()
+            .trustStore((String) serverContext.getServerConfig().get(ConfigParamNames.TRUST_STORE))
+            .tsPasswordFile((String) serverContext.getServerConfig().get(ConfigParamNames.TRUST_STORE_PASS_FILE))
+            .keyStore((String) serverContext.getServerConfig().get(ConfigParamNames.KEY_STORE))
+            .ksPasswordFile((String) serverContext.getServerConfig().get(ConfigParamNames.KEY_STORE_PASS_FILE))
+            .tlsEnabled((Boolean) serverContext.getServerConfig().get("--enable-tls"))
+            .systemDownHandler(() -> System.exit(SYSTEM_EXIT_ERROR_CODE))
+            .maxCacheEntries(serverContext.getLogReplicationCacheMaxSize()/2)
+            .maxWriteSize(serverContext.getMaxWriteSize())
+            .build())
+            .parseConfigurationString(localCorfuEndpoint).connect();
+    }
+
+    private String getCorfuEndpoint(String localHostAddress, int port) {
+        return getVersionFormattedEndpointURL(localHostAddress, port);
+    }
+
+    private boolean isUpgradeInProgress(LogReplicationUpgradeManager upgradeManager, CorfuStore corfuStore) {
```

Comment:
Since startDiscoveryService polls on this method, do you think we can use notification or some other mechanism, rather than reading from DB every time?

Reply:
Yes, good suggestion.  Let me think if something we can add something like that.